api/coreservice.go

```diff
@@ -1377,26 +1383,62 @@ func (core *coreService) LogsInRange(filter *logfilter.LogFilter, start, end, pa
 		return nil, err
 	}
 
-	// TODO: improve using goroutine
 	if paginationSize == 0 {
 		paginationSize = 1000
 	}
 	if paginationSize > 5000 {
 		paginationSize = 5000
 	}
+
 	logs := []*iotextypes.Log{}
+	jobs := make(chan uint64, len(blockNumbers))
+	results := make(chan _logsInBlock, len(blockNumbers))
+	var wg sync.WaitGroup
+	if len(blockNumbers) == 0 {
+		return logs, nil
+	}
 	for _, i := range blockNumbers {
-		logsInBlock, err := core.LogsInBlock(filter, i)
-		if err != nil {
+		jobs <- i
+	}
+	close(jobs)
+
+	for w := 1; w <= 5; w++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for job := range jobs {
+				logsInBlock, err := core.logsInBlock(filter, job)
+				if err != nil {
+					results <- _logsInBlock{
+						nil,
+						err,
+					}
+					return
+				}
+
+				results <- _logsInBlock{
+					logsInBlock,
+					nil,
+				}
+			}
+		}()
+	}
+	wg.Wait()
+
+	for i := 0; i < len(blockNumbers); i++ {
+		logsInBlock := <-results
```

Comment:
this is not in order. build a map[uint64]logs first, and then iterate by block number